File: runtime-go/domain/event_payloads.go
package domain

// EventPayload 是所有 Event Payload 的 marker 接口。
// 每个具体 payload 类型实现 eventPayload()，只是为了让 Event.Payload
// 在类型系统上收紧一层——避免任何东西都能塞进 Payload。
type EventPayload interface {
	eventPayload()
}

// ---------- Session ----------

type PayloadSessionOpened struct {
	Principal string
	Metadata  map[string]string
}

func (PayloadSessionOpened) eventPayload() {}

// ---------- Task ----------

type PayloadTaskCreated struct {
	Goal     string
	Budget   Budget
	ParentID string // 空 = 根 Task；与 Task.ParentID 对齐(ch07)
}

func (PayloadTaskCreated) eventPayload() {}

type PayloadTaskEnded struct {
	Status TaskStatus
	Reason string // 可选：canceled/timeout/failed 时说明原因
}

func (PayloadTaskEnded) eventPayload() {}

// ---------- Turn ----------

type PayloadTurnStarted struct {
	Index int
}

func (PayloadTurnStarted) eventPayload() {}

type PayloadTurnEnded struct {
	Status    TurnStatus
	TokensIn  int
	TokensOut int
	CostUS    float64
	LatencyMS int64
}

func (PayloadTurnEnded) eventPayload() {}

// ---------- 用户输入 ----------

type PayloadUserSpoke struct {
	Text string
}

func (PayloadUserSpoke) eventPayload() {}

// ---------- LLM 交互 ----------

type PayloadLLMRequested struct {
	Model    string
	Messages []Message // 组装好的最终上下文
	Tools    []Tool
}

func (PayloadLLMRequested) eventPayload() {}

type PayloadLLMReplied struct {
	Assistant Message
	ToolCalls []ToolCall
	TokensIn  int
	TokensOut int
}

func (PayloadLLMReplied) eventPayload() {}

// ---------- 工具调用 ----------

type PayloadToolCalled struct {
	CallID    string
	Name      string
	Arguments string // JSON 字符串,与 ToolCall.Arguments 一致
}

func (PayloadToolCalled) eventPayload() {}

type PayloadToolReturned struct {
	CallID  string
	Content string // 序列化后的返回值
	IsError bool
}

func (PayloadToolReturned) eventPayload() {}

// ---------- 上下文压缩 (ch04) ----------

// PayloadContextCompressed 说明"哪一段 Event 被压缩成什么摘要"。见 ch04 §4.5.3。
// 是回放性的核心保证:摘要作为不可变事实进 EventStore。
type PayloadContextCompressed struct {
	FromSeq    int64   // 覆盖的 seq 范围起点
	ToSeq      int64   // 覆盖的 seq 范围终点
	Strategy   string  // "turns:8" | "task-end" | "manual" | "fallback:flat" | ...
	Summary    Summary // 结构化摘要,见 §4.6.1
	FromTokens int     // 压缩前估算 token(可选)
	ToTokens   int     // 摘要后 token(可选)
}

func (PayloadContextCompressed) eventPayload() {}

// PayloadCompressionSkipped 记录"这次压缩尝试失败/跳过"。见 ch04 §4.9。
type PayloadCompressionSkipped struct {
	Reason string // "llm_error" | "schema_invalid" | ...
	Detail string
}

func (PayloadCompressionSkipped) eventPayload() {}

// ---------- 任务进度 (ch04 §4.7) ----------

// PayloadProgressUpdated 记录一次 Progress 折叠。幂等:同 Version 写两次结果一样。
type PayloadProgressUpdated struct {
	TaskID   string
	Progress Progress
}

func (PayloadProgressUpdated) eventPayload() {}

// ---------- Memory (ch05 会展开) ----------

// PayloadMemoryQueried 记录一次向量库/KV 检索的结果。ch04 只定义,ch05 用。
type PayloadMemoryQueried struct {
	Query string
	Refs  []MemoryRef
}

func (PayloadMemoryQueried) eventPayload() {}

// ---------- Task Graph (ch07) ----------

// PayloadSubTaskSpawned 记录 Planner 为父 Task 派生出一个子 Task。
// 与 TaskCreated 的关系:SubTaskSpawned 是 Planner 的意图事件;
// 协调器/Loop 随后追加 TaskCreated{ParentID} 真正创建子 Task。
// Round 2 参考实现里 Plan 直接产出 TaskCreated(带 ParentID),本 payload 保留给显式图编排。
type PayloadSubTaskSpawned struct {
	ParentTaskID string
	ChildTaskID  string
	Goal         string
	Budget       Budget
}

func (PayloadSubTaskSpawned) eventPayload() {}

// ---------- Tool binding (ch08) ----------

// PayloadToolBindFailed 记录工具注册表里找不到或 schema 校验失败。
type PayloadToolBindFailed struct {
	CallID string
	Name   string
	Reason string // "unknown_tool" | "schema_invalid" | ...
}

func (PayloadToolBindFailed) eventPayload() {}

File: runtime-rs/src/domain/event_payloads.rs
//! Event Payload 的类型系统。
//!
//! Go 版用 marker interface 收紧 `Event.Payload`；Rust 天然的对应物是
//! 一个封闭的 enum —— 编译器强制穷举，比 marker trait 更严格。
//!
//! 每个 variant 对应 Go 版一个 EventType，字段与 Go 端 Payload* struct 对齐。

use serde::{Deserialize, Serialize};

use super::summary::{MemoryRef, Progress, Summary};
use super::{Budget, Message, TaskStatus, Tool, ToolCall, TurnStatus};

/// EventType 与 Payload 的 discriminant 合并成一个类型。
/// 判别一个 Event 的类型 = match 它的 Payload。
pub type EventType = &'static str;

pub const EVT_SESSION_OPENED: EventType = "SessionOpened";
pub const EVT_TASK_CREATED: EventType = "TaskCreated";
pub const EVT_TASK_ENDED: EventType = "TaskEnded";
pub const EVT_TURN_STARTED: EventType = "TurnStarted";
pub const EVT_TURN_ENDED: EventType = "TurnEnded";
pub const EVT_USER_SPOKE: EventType = "UserSpoke";
pub const EVT_LLM_REQUESTED: EventType = "LLMRequested";
pub const EVT_LLM_REPLIED: EventType = "LLMReplied";
pub const EVT_TOOL_CALLED: EventType = "ToolCalled";
pub const EVT_TOOL_RETURNED: EventType = "ToolReturned";
pub const EVT_CONTEXT_COMPRESSED: EventType = "ContextCompressed";
pub const EVT_COMPRESSION_SKIPPED: EventType = "CompressionSkipped";
pub const EVT_PROGRESS_UPDATED: EventType = "ProgressUpdated";
pub const EVT_MEMORY_QUERIED: EventType = "MemoryQueried";

/// 与 Go 版 `EventPayload` marker interface 对等。
/// 在 Rust 里用封闭 enum：外部无法新增 variant，消费方 match 必须穷举。
///
/// `#[serde(tag="type", content="payload")]` 让 wire format 与 Go 端 `EventDTO`
/// 一致:`{"type":"UserSpoke","payload":{...}}`。见 ch03 §3.3.2。
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", content = "payload")]
pub enum EventPayload {
    SessionOpened(PayloadSessionOpened),
    TaskCreated(PayloadTaskCreated),
    TaskEnded(PayloadTaskEnded),
    TurnStarted(PayloadTurnStarted),
    TurnEnded(PayloadTurnEnded),
    UserSpoke(PayloadUserSpoke),
    LLMRequested(PayloadLLMRequested),
    LLMReplied(PayloadLLMReplied),
    ToolCalled(PayloadToolCalled),
    ToolReturned(PayloadToolReturned),
    ContextCompressed(PayloadContextCompressed),
    CompressionSkipped(PayloadCompressionSkipped),
    ProgressUpdated(PayloadProgressUpdated),
    MemoryQueried(PayloadMemoryQueried),
}

impl EventPayload {
    pub fn event_type(&self) -> EventType {
        match self {
            EventPayload::SessionOpened(_)     => EVT_SESSION_OPENED,
            EventPayload::TaskCreated(_)       => EVT_TASK_CREATED,
            EventPayload::TaskEnded(_)         => EVT_TASK_ENDED,
            EventPayload::TurnStarted(_)       => EVT_TURN_STARTED,
            EventPayload::TurnEnded(_)         => EVT_TURN_ENDED,
            EventPayload::UserSpoke(_)         => EVT_USER_SPOKE,
            EventPayload::LLMRequested(_)      => EVT_LLM_REQUESTED,
            EventPayload::LLMReplied(_)        => EVT_LLM_REPLIED,
            EventPayload::ToolCalled(_)        => EVT_TOOL_CALLED,
            EventPayload::ToolReturned(_)      => EVT_TOOL_RETURNED,
            EventPayload::ContextCompressed(_) => EVT_CONTEXT_COMPRESSED,
            EventPayload::CompressionSkipped(_) => EVT_COMPRESSION_SKIPPED,
            EventPayload::ProgressUpdated(_) => EVT_PROGRESS_UPDATED,
            EventPayload::MemoryQueried(_) => EVT_MEMORY_QUERIED,
        }
    }
}

// ---------- 具体 payload 结构 ----------

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct PayloadSessionOpened {
    pub principal: String,
    pub metadata: std::collections::HashMap<String, String>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct PayloadTaskCreated {
    pub goal: String,
    pub budget: Budget,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct PayloadTaskEnded {
    pub status: TaskStatus,
    pub reason: String,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct PayloadTurnStarted {
    pub index: i32,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct PayloadTurnEnded {
    pub status: TurnStatus,
    pub tokens_in: i64,
    pub tokens_out: i64,
    pub cost_us: f64,
    pub latency_ms: i64,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct PayloadUserSpoke {
    pub text: String,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct PayloadLLMRequested {
    pub model: String,
    pub messages: Vec<Message>,
    pub tools: Vec<Tool>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct PayloadLLMReplied {
    pub assistant: Message,
    pub tool_calls: Vec<ToolCall>,
    pub tokens_in: i64,
    pub tokens_out: i64,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct PayloadToolCalled {
    pub call_id: String,
    pub name: String,
    pub arguments: String,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct PayloadToolReturned {
    pub call_id: String,
    pub content: String,
    pub is_error: bool,
}

/// 与 Go 版 `PayloadContextCompressed` 对齐。见 ch04 §4.5.3。
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct PayloadContextCompressed {
    pub from_seq: i64,
    pub to_seq: i64,
    pub strategy: String,
    pub summary: Summary,
    pub from_tokens: i64,
    pub to_tokens: i64,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct PayloadCompressionSkipped {
    pub reason: String,
    pub detail: String,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct PayloadProgressUpdated {
    pub task_id: String,
    pub progress: Progress,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct PayloadMemoryQueried {
    pub query: String,
    pub refs: Vec<MemoryRef>,
}
